src/core/thread/mesh_forwarder.cpp

```diff
@@ -206,6 +214,168 @@ void MeshForwarder::ScheduleTransmissionTask()
+            mSrcMatchEnabled = false;
+        }
+        else
+        {
+            // succeed in adding to source match table
+            aChild.mAddSrcMatchEntryPending = false;
+        }
+    }
+
+exit:
+
+    if (!mSrcMatchEnabled)
+    {
+        if (aChild.mAddSrcMatchEntryPending && (aChild.mQueuedIndirectMessageCnt == 1))
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchOverflowShortCnt++;
+            }
+            else
+            {
+                mSrcMatchOverflowExtCnt++;
+            }
+
+        }
+
+        otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                     mSrcMatchOverflowShortCnt,
+                     mSrcMatchOverflowExtCnt,
+                     mSrcMatchReleasedShortCnt,
+                     mSrcMatchReleasedExtCnt);
+    }
+
+    return;
+}
+
+void MeshForwarder::ClearSrcMatchEntry(Child &aChild)
+{
+    Mac::Address macAddr;
+
+    if (aChild.mAddSrcMatchEntryShort)
+    {
+        macAddr.mLength = sizeof(macAddr.mShortAddress);
+        macAddr.mShortAddress = aChild.mValid.mRloc16;
+    }
+    else
+    {
+        macAddr.mLength = sizeof(macAddr.mExtAddress);
+        memcpy(macAddr.mExtAddress.m8, aChild.mMacAddr.m8, sizeof(macAddr.mExtAddress));
+    }
+
+    if (mMac.ClearSrcMatchEntry(macAddr) == kThreadError_None)
+    {
+        // if waiting for more source table resources
+        if (!mSrcMatchEnabled)
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchReleasedShortCnt++;
+            }
+            else
+            {
+                mSrcMatchReleasedExtCnt++;
+            }
+        }
+    }
+    else
+    {
+        // if finished queued messages for SED which is not added into the source match table
+        if (aChild.mAddSrcMatchEntryShort)
+        {
+            mSrcMatchOverflowShortCnt--;
+        }
+        else
+        {
+            mSrcMatchOverflowExtCnt--;
+        }
+
+        aChild.mAddSrcMatchEntryPending = false;
+    }
+
+    otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                 mSrcMatchOverflowShortCnt,
+                 mSrcMatchOverflowExtCnt,
+                 mSrcMatchReleasedShortCnt,
+                 mSrcMatchReleasedExtCnt);
+
+    // use tasklet for re-enable source match to avoid possible disorder due to
+    // possible repeatedly disable/enable when many SEDs exists
+    mUpdateSrcMatchTask.Post();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask(void *aContext)
+{
+    MeshForwarder *obj = reinterpret_cast<MeshForwarder *>(aContext);
+    obj->HandleUpdateSrcMatchTask();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask()
+{
+    // if total available source match table resources more than total required resources
+    if (!mSrcMatchEnabled &&
+        (mSrcMatchReleasedShortCnt + 2 * mSrcMatchReleasedExtCnt) >= (mSrcMatchOverflowShortCnt + 2 * mSrcMatchOverflowExtCnt))
```

Comment:
I don't quite understand this overflow logic.  I would expect this logic to make sure that neither the short nor extended address tables are overflowing.  Remember, the short and extended address tables may be completely separate for some radios.  As a result, I would expect to see a check like:

```
    // if total available source match table resources more than total required resources
    if (!mSrcMatchEnabled &&
        (mSrcMatchReleasedShortCnt >= mSrcMatchOverflowShortCnt) &&
        (mSrcMatchReleasedExtCnt >= mSrcMatchOverflowExtCnt))
```